open-apparel-registry/open-apparel-registry
# Add models to persist embed config

## Overview

The embed config is complex and will grow over time as we add features. Rather than serialize to the query string, store the config in the database and link it to the contributor. 
- Adds configuration models
- Adds creation, update and fetching of embedded map configurations via API
     - Note that the nested embedded map fields are deleted and recreated on update of the parent `EmbedConfig`
- Adds embedded map configuration to the data fetched for user profiles

Connects #1299 

## Demo

<img width="1179" alt="Screen Shot 2021-04-06 at 11 58 20 AM" src="https://user-images.githubusercontent.com/21046714/113916496-08a3fa80-97ae-11eb-93b3-ac76997aa2db.png">

## Testing Instructions

* Run `./scripts/manage migrate`
* Run `./scripts/server` and login as a user with an associated contributor
* Navigate to [:8081/api/embed-configs/](http://localhost:8081/api/embed-configs/)
* `POST` a new embed config object
     - [x] The new EmbedConfig should be created successfully
```
     {
	"width": "100",
	"height": "100",
	"color": "pink",
	"font": "serif",
	"show_other_contributor_information": false,
	"embed_fields": [{
		"column_name": "xyz",
		"display_name": "123",
		"visible": true,
		"order": 1
	}]
   }
```
* Navigate to [your profile](http://localhost:6543/profile/8]
     - [ ] Confirm in the network tab that the EmbedConfig is returned with the contributor details
* Navigate to [your new EmbedConfig](http://localhost:8081/api/embed-configs/1/) and update some fields, including the embed_fields
     - [x] The configuration should update successfully
* Attempt to create an embed_field with a duplicate order
     - [x] You should receive an error message
* Attempt to create a new EmbedConfig with the same user
     - [ ] You should receive an error message

## Checklist
